[WIP] Patch for media response context explosion

shaunak99 opened the pull request:
This patch addresses a bug where toolResponses containing media type data as base64 strings leads to exploded contexts due to addition into the context window as text parts. Currently, we implement a defensive approach and return media types for UI rendering while filtering it completely in the context.

Ideally, the media type should be parsed and filtered based on the model capabilities and added to the context with appropriate text/image/audio/file part. 

Note: Only OpenAI/Anthropic services are addressed in this since Vercel automatically re-ingests toolResponses and needs some explicit work around to prevent this. - Needs to be addressed asap or else token cost can explode very easily.

rahulkarajgikar commented on tts_Hello_20250802_025327.mp3:
delete this?

rahulkarajgikar commented on src/core/llm/services/vercel.ts:
```diff
@@ -38,6 +39,8 @@ export class VercelLLMService implements ILLMService {
     private readonly sessionId: string;
     private toolSupportCache: Map<string, boolean> = new Map();
+    // Map of toolCallId -> queue of raw results to emit with that callId
+    private rawResultsByCallId: Map<string, any[]> = new Map();
```
use unknown not any

or better, get the RESULT type mentioned here: https://ai-sdk.dev/docs/reference/ai-sdk-core/stream-text#tools.tool.execute

rahulkarajgikar commented on src/core/llm/services/vercel.ts:
```diff
@@ -92,14 +95,53 @@ export class VercelLLMService implements ILLMService {
                 acc[toolName] = {
                     inputSchema: jsonSchema(tool.parameters),
-                    execute: async (args: unknown) => {
+                    execute: async (args: unknown, _options?: { toolCallId?: string }) => {
```
don't use underscore for used variables

rahulkarajgikar commented on src/core/llm/services/vercel.ts:
```diff
@@ -92,14 +95,53 @@ export class VercelLLMService implements ILLMService {
                             );
+
+                            // Queue raw, unfiltered result under the specific toolCallId
+                            const callId = _options?.toolCallId ?? this.sessionId;
+                            const queue = this.rawResultsByCallId.get(callId) ?? [];
```
don't fallback call id to session id, make the toolCallId non optional (vercel always provides it)

and use that

https://ai-sdk.dev/docs/reference/ai-sdk-core/stream-text

rahulkarajgikar commented on src/core/llm/services/vercel.ts:
```diff
@@ -92,14 +95,53 @@ export class VercelLLMService implements ILLMService {
+                                const safeContent = sanitizeToolResultToContent(rawResult);
+                                const summaryText = summarizeToolContentForText(safeContent);
+                                return summaryText;
+                            } catch (sanErr) {
+                                // Fallback: stringify/truncate to keep context safe
+                                const asString =
+                                    typeof rawResult === 'string'
+                                        ? rawResult
+                                        : (() => {
+                                              try {
+                                                  return JSON.stringify(rawResult);
+                                              } catch {
+                                                  return String(rawResult);
+                                              }
+                                          })();
+                                const MAX_LEN = 4000;
+                                const truncated =
+                                    asString.length > MAX_LEN
+                                        ? `${asString.slice(0, MAX_LEN)}... [truncated]`
+                                        : asString;
+                                logger.warn(
+                                    `Vercel tool result sanitization failed for ${toolName}: ${String(
+                                        sanErr
+                                    )}. Returning truncated text.`
+                                );
+                                return truncated;
+                            }
```
we already handle all of this error stuff in the helper functions so we don't need this catch

`sanitizeToolResultToContent` is already a safe function that catches errors, and handles them safely like this, so it never errors out so we don't need this

we can make the summarize func also safe in simialr manner

rahulkarajgikar commented on src/core/llm/services/vercel.ts:
```diff
@@ -298,12 +339,22 @@ export class VercelLLMService implements ILLMService {
                         for (const toolResult of step.toolResults) {
+                            const callId = toolResult.toolCallId;
+                            const sanitized = toolResult.output;
+                            let raw: any | undefined;
```
don't use any type, similar to above (use REsult type or unknown)

rahulkarajgikar commented on src/core/context/manager.ts:
```diff
@@ -432,30 +432,30 @@ export class ContextManager<TMessage = unknown> {
-        } else {
-            // Fallback: stringify all other values
-            content = JSON.stringify(result ?? '');
+        content = sanitizeToolResultToContent(result);
```
currently vercel vs anthropic/openai handle getting llm responses quite differently

(anthropic/openai add the tool responses, assistant resposnes using the individual methods, while vercel uses processLLMResponse, processStreamResponse.) - let's add A TODO at the top of context manager file to unify these approaches so it's more uniform everywhere and easier to make changes in 1 place for all

rahulkarajgikar commented on src/core/context/utils.ts:
```diff
@@ -206,7 +206,295 @@ export function filterMessagesByLLMCapabilities(
+ * - Removes huge binary blobs inside objects
+ * - Truncates extremely long raw text
+ */
+export function sanitizeToolResultToContent(result: unknown): InternalMessage['content'] {
```
add unit tests for this function for known result types to validate it works properly

rahulkarajgikar commented on src/core/context/utils.ts:
```diff
@@ -206,7 +206,295 @@ export function filterMessagesByLLMCapabilities(
+        if (Array.isArray(result)) {
+            // Ensure only supported part types (text|image|file) appear in the array
+            const parts: Array<TextPart | ImagePart | FilePart> = [];
+            for (const item of result as any[]) {
```
don't use any, use unknown and add checks for types if needed

rahulkarajgikar commented on src/core/context/utils.ts:
```diff
@@ -206,7 +206,295 @@ export function filterMessagesByLLMCapabilities(
+                        });
+                        continue;
+                    }
+                    if ('image' in obj) {
+                        parts.push({
+                            type: 'image',
+                            image: getImageData({ image: obj.image }),
+                            mimeType: obj.mimeType || 'image/jpeg',
+                        });
+                        continue;
```
can combine with prev if

rahulkarajgikar commented on src/core/context/utils.ts:
```diff
@@ -206,7 +206,295 @@ export function filterMessagesByLLMCapabilities(
+ * - If content is a string that looks like base64/data URI, replace with a short placeholder.
+ * - Otherwise pass text through.
+ */
+export function toTextForToolMessage(content: InternalMessage['content'] | string | null): string {
```
`| string | null` are already covered in `InternalMessage['content']` so this is redundant, we can just directly use `content: InternalMessage['content']`

rahulkarajgikar commented on src/core/context/utils.ts:
```diff
@@ -206,7 +206,295 @@ export function filterMessagesByLLMCapabilities(
+/**
+ * Parse data URI and return { mediaType, base64 } or null if not a data URI.
+ */
+export function parseDataUri(value: string): { mediaType: string; base64: string } | null {
```
add few unit tests to validate base64 truncation/etc. properly

rahulkarajgikar reviewed (commented):
alright, let's just test out once and see if the base64 data is flooding UI. 

if yeah, we can temporarily do coderabbit's changes to avoid this, and do a better fix later 

otherwise this is cool to merge

shaunak99 commented:
So we do want the UI to get the raw base64 - this is how the UI renders the image/audio

However, logging the raw was definitely problematic so I have updated that. Works as expected.

Also updated `copy-webui-dist.ts` to patch the webui build issue
